File: server/src/test/java/com/pass/unit/exam/domain/ExamTitleTest.java
package com.pass.unit.exam.domain;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import com.pass.exam.domain.ExamTitle;
import com.pass.global.exception.BadRequestException;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

class ExamTitleTest {

    @Test
    @DisplayName("시험 제목은 생성할 수 있다.")
    void create() {
        // given
        String title = "a".repeat(100);

        // when
        ExamTitle examTitle = new ExamTitle(title);

        // then
        assertThat(examTitle.getValue()).isEqualTo(title);
    }

    @Test
    @DisplayName("시험 제목은 비어있을 수 없다.")
    void titleNotNullAndNotBlank() {
        assertThatThrownBy(() -> new ExamTitle(null))
                .isInstanceOf(BadRequestException.class)
                .hasMessage("시험 제목은 비어있을 수 없습니다.");
        assertThatThrownBy(() -> new ExamTitle(" "))
                .isInstanceOf(BadRequestException.class)
                .hasMessage("시험 제목은 비어있을 수 없습니다.");
    }

    @Test
    @DisplayName("시험 제목은 100자 이하여야 한다.")
    void titleLength() {
        // given
        String title = "a".repeat(101);

        // when & then
        assertThatThrownBy(() -> new ExamTitle(title))
                .isInstanceOf(BadRequestException.class)
                .hasMessage("시험 제목은 100자 이하여야 합니다.");
    }
}

File: server/src/test/kotlin/com/fluffy/exam/domain/ExamTitleTest.kt
package com.fluffy.exam.domain

import com.fluffy.global.exception.BadRequestException
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.StringSpec
import io.kotest.data.forAll
import io.kotest.data.row
import io.kotest.matchers.shouldBe

class ExamTitleTest : StringSpec({

    "시험 제목이 정상적으로 생성된다." {
        // given
        val title = "a".repeat(100)

        // when
        val examTitle = ExamTitle(title)

        // then
        examTitle.value shouldBe title
    }

    "시험 제목은 비어있을 수 없다." {
        forAll(
            row(""),
            row(" ")
        ) {
            shouldThrow<BadRequestException> { ExamTitle(it) }
                .message shouldBe "시험 제목은 비어있을 수 없습니다."
        }
    }

    "시험 제목은 100자 이하여야 한다." {
        // given
        val title = "a".repeat(101)

        // when & then
        shouldThrow<BadRequestException> { ExamTitle(title) }
            .message shouldBe "시험 제목은 100자 이하여야 합니다."
    }
})

